src/transports/ws.rs

```diff
@@ -1,186 +1,309 @@
-                                                log::warn!("Got notification for unknown subscription (id: {:?})", id);
-                                            }
-                                        } else {
-                                            log::error!("Got unsupported notification (id: {:?})", id);
-                                        }
-                                    }
-
-                                    return Ok(());
-                                }
-
-                                let response = helpers::to_response_from_slice(t.as_bytes());
-                                let outputs = match response {
-                                    Ok(rpc::Response::Single(output)) => vec![output],
-                                    Ok(rpc::Response::Batch(outputs)) => outputs,
-                                    _ => vec![],
-                                };
-
-                                let id = match outputs.get(0) {
-                                    Some(&rpc::Output::Success(ref success)) => success.id.clone(),
-                                    Some(&rpc::Output::Failure(ref failure)) => failure.id.clone(),
-                                    None => rpc::Id::Num(0),
-                                };
-
-                                if let rpc::Id::Num(num) = id {
-                                    if let Some(request) = pending_.lock().remove(&(num as usize)) {
-                                        log::trace!("Responding to (id: {:?}) with {:?}", num, outputs);
-                                        if let Err(err) = request.send(helpers::to_results_from_outputs(outputs)) {
-                                            log::warn!("Sending a response to deallocated channel: {:?}", err);
-                                        }
-                                    } else {
-                                        log::warn!("Got response for unknown request (id: {:?})", num);
-                                    }
-                                } else {
-                                    log::warn!("Got unsupported response (id: {:?})", id);
-                                }
-
-                                Ok(())
-                            }
-                            _ => Ok(()),
+    async fn into_task(self, requests: mpsc::UnboundedReceiver<TransportMessage>) {
+        let Self {
+            receiver,
+            mut sender,
+            mut pending,
+            mut subscriptions,
+        } = self;
+
+        let receiver = connection::into_stream(receiver);
+        let receiver = receiver.fuse();
+        let requests = requests.fuse();
+        pin_mut!(receiver);
+        pin_mut!(requests);
+        loop {
+            select! {
+                msg = requests.next() => match msg {
+                    Some(TransportMessage::Request { id, request, sender: tx }) => {
+                        if pending.insert(id.clone(), tx).is_some() {
```

Comment:
I've decided to rather `remove` when it's unsuccessful, otherwise we risk a race between getting a response. Although on the second thought, since we use `select` the whole future can only progress in a single thread, so probably it won't happen, but I'd rather be safe than sorry.